src/TorchSharp/JIT/ScriptModule.cs

```diff
@@ -0,0 +1,338 @@
+                        case Modules.Parameter param when dtype == param.dtype:
+                            continue;
+
+                        case Modules.Parameter param: {
+                                var t = param.to(dtype);
+                                t.retain_grad();
+                                var p = new Modules.Parameter(t, param.requires_grad);
+                                field.SetValue(this, p);
+                                ConditionallyRegisterParameter(fieldName, p);
+                                break;
+                            }
+
+                        case Tensor tensor when dtype == tensor.dtype:
+                            continue;
+
+                        case Tensor tensor: {
+                                var t = tensor.to(dtype);
+                                field.SetValue(this, t);
+                                ConditionallyRegisterBuffer(fieldName, t);
+                                break;
+                            }
+                        }
+                    }
+
+                    return this;
+                }
+
+#if false   // These functions "work," but the native code doesn't seem to find any interesting information.
+
+                [DllImport("LibTorchSharp")]
+                private static extern IntPtr THSJIT_Module_getInputType(HType module, int index);
+
+                public Type GetInputType(int index)
+                {
+                    var type = new Type(THSJIT_Module_getInputType(handle, index), Type.TypeKind.AnyType);
+
+                    return GetType(type);
+                }
+
+                [DllImport("LibTorchSharp")]
+                private static extern IntPtr THSJIT_getOutputType(HType module, int index);
+
+                public Type GetOutputType(int index)
+                {
+                    var type = new Type(THSJIT_getOutputType(handle, index), Type.TypeKind.AnyType);
+
+                    return GetType(type);
+                }
+
+                private Type GetType(Type type)
+                {
+                    switch (type.Kind) {
+                    case Type.TypeKind.TensorType:
+                        var dynamic = type.AsTensorType();
+                        type.Dispose();
+                        return dynamic;
+                    //case Type.TypeKind.DimensionedTensorType:
+                    //    var tensor = type.AsTensorType();
+                    //    type.Dispose();
+                    //    return tensor;
+                    default:
+                        return type;
+                    }
+                }
+#endif
+
+                [DllImport("LibTorchSharp")]
+                private static extern IntPtr THSJIT_Module_forward(HType module, IntPtr tensors, int length);
+
+                public unsafe override Tensor forward(Tensor tensor)
+                {
+                    using (var parray = new PinnedArray<IntPtr>()) {
```

Comment:
@tarekgh -- is there some way to avoid the new[]{} and use stackalloc here, instead?

Reply:
You don't need to use PinnedArray at all too. You may try the following which will allow writing the whole method without any unsafe code and you'll not to do any managed allocations.

```C#
ReadOnlySpan<IntPtr> buffer = stackalloc IntPtr[1] { tensor.Handle };
var res = THSJIT_Module_forward(handle, (IntPtr)tensorRefs, 1);
```